Feature/ethernet support

mobrembski opened the pull request:
Basic support of Ethernet in ESP32
Based on Olimex ESP32-POE board

Currently, only DHCP mode is supported, no static IP.
But at least it works along with WiFi.

mobrembski committed:
Initial Ethernet support

mobrembski committed:
Added displaying Ethernet parameter in sysinfo and root page

TD-er commented on src/ESPEasyEth.ino:
```diff
@@ -0,0 +1,35 @@
+bool prepareEth() {
+  char hostname[40];
+  safe_strncpy(hostname, EthGetHostname(), sizeof(hostname));
+  ETH.setHostname(hostname);
```
Why not use `ETH.setHostname(EthGetHostname().c_str());` ?

mobrembski replied:
:D 
Right, idiotic....

TD-er commented:
I was expecting some code changes also in the check whether we're connected.
Is that all handled in the events?

And do you have a direct link to the ETH module you're using?

mobrembski commented:
> I was expecting some code changes also in the check whether we're connected.
> Is that all handled in the events?
> 
> And do you have a direct link to the ETH module you're using?

I have an Olimex ESP32-PoE board:
[Olimex ESP32-PoE](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware)
It uses LAN8710A Phy chip.

I haven't added any code for handling event, because i'm not sure right now if they're required.
Why? Because everything is done automatically by espressif framework.
They deal with ESP32 Eth interrupts i guess, because only ```ETH.begin()``` with proper configuration is needed to establish connection. You can disconnect/reconnect cable, and it will get IP.
Moreover, it uses same TCP/IP stack as WiFi, so there's even no restart of listening socket required.

Right now i'm trying to include Eth Phy configuration (MDIO,MCS, Power GPIO selection, Phy type selection etc.) into ESPEasy settings, but i cant find place which is executed after submit is pressed on Settings page?

uzi18 commented:
@TD-er eth module type can be configured before initialization/begin, but here is used as static.
We can save configuration in another file like: pins configuration and ethernet type, IP configuration etc.

TD-er commented:
When the submit button is pressed, you effectively do a reload of that page with some POST data.
So all handle functions for the web pages have a section where data is fetched from the webserver (`.arg` function) and stored in the settings.
The second part of those handle functions is to render the page.

In ESPEasy we do check a lot for the wifi connected state, which is done in a function that's called a lot.
In the WiFi events we have a number of events like when a connection is established and also when an IP-addres is received from the DHCP server.

Also we have some places where we directly address the WiFi object (not a lot places) so those must be done also for the ETH object.

mobrembski committed:
Added configuration of Eth Phy via Hardware settings web page

uzi18 commented on src/ESPEasy_checks.ino:
```diff
@@ -34,7 +34,7 @@ template<typename T, typename U> constexpr size_t offsetOf(U T::*member)
   check_size<CRCStruct,                             168u>();
   check_size<SecurityStruct,                        593u>();
-  const unsigned int SettingsStructSize = (244 + 82 * TASKS_MAX);
+  const unsigned int SettingsStructSize = (252 + 82 * TASKS_MAX);
```
244+6 ? 
should be conditional with  HAS_ETHERNET

mobrembski replied:
Actually it passes after 252, not 260 :/
But yes, you've got right it shou;d be under HAS_ETHERNET !

TD-er commented on src/src/DataStructs/SettingsStruct.h:
```diff
@@ -175,6 +175,14 @@ class SettingsStruct_tmpl
+  int8_t        ETH_Pin_mdio;
+  int8_t        ETH_Pin_power;
+  int8_t        ETH_Phy_Type;
+  uint8_t       ETH_Clock_Mode;
```
Make sure these values do get some proper initialization values.
This means for existing settings you need to initialize them to proper defaults (e.g. 0)
and increase the internal BUILD value (is now 20105) and also init it to proper values when initializing new settings.
Would be nice if you can fit these values in a 32 bit value, so we don't get padding of the settings.
A 32-bit value is always 4-byte aligned, but 8-bit values are not.
You can then make a get/set function in the struct to handle these parameters.

mobrembski replied:
Hmm i can do some bit-manipulations to use only one uint32_t, if i understood you correctly.

I will also add ESP32-PoE to DeviceModel, so it could be possible to set pre-defined configuration.

mobrembski replied:
I'm not really sure why in this case it's better to put this values in 32bit?
As i can see, there are a lot of uint8_t or 16t or even booleans in SettingsStruct right now

TD-er replied:
For existing members you must not change their signature as it may also affect their position in the struct.
For new members you will add at the end, you determine where the next member will be positioned.

A "large" member will be positioned in memory at an address which is a multiple of its own size (max modulo 4 bytes on this platform)
For example, a struct with these members:
```
char
uint32_t
char
uint32_t
```
will be 16 bytes, while it can be 10 bytes if you order them by size, largest first.

uzi18 replied:
you don't need to change anything here, @TD-er just explained how align work on esp

mobrembski replied:
Okay, i know about struct alignment in C, but i just couldn't find reason how this could be inefficient in case of 6 8-bit values ;)

TD-er replied:
You don't need 6x 8 bits, so you could store it in 32 bit :)

But maybe it is a bit too much effort for at most 2 bytes gain in settings struct size.
Just leave it.

ironataerial commented:
> > I was expecting some code changes also in the check whether we're connected.
> > Is that all handled in the events?
> > And do you have a direct link to the ETH module you're using?
> 
> I have an Olimex ESP32-PoE board:
> [Olimex ESP32-PoE](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware)
> It uses LAN8710A Phy chip.
> 
> I haven't added any code for handling event, because i'm not sure right now if they're required.
> Why? Because everything is done automatically by espressif framework.
> They deal with ESP32 Eth interrupts i guess, because only `ETH.begin()` with proper configuration is needed to establish connection. You can disconnect/reconnect cable, and it will get IP.
> Moreover, it uses same TCP/IP stack as WiFi, so there's even no restart of listening socket required.
> 
> Right now i'm trying to include Eth Phy configuration (MDIO,MCS, Power GPIO selection, Phy type selection etc.) into ESPEasy settings, but i cant find place which is executed after submit is pressed on Settings page?

From Olimex product page :

",,,The Si3402-B chip is IEEE 802.3-compliant, including pre-standard (legacy) PoE support. ..."

Can you confirm the PoE part operates on the non standard (aka passive) PoE 12-48v ?

mobrembski commented:
> > > I was expecting some code changes also in the check whether we're connected.
> > > Is that all handled in the events?
> > > And do you have a direct link to the ETH module you're using?
> > 
> > 
> > I have an Olimex ESP32-PoE board:
> > [Olimex ESP32-PoE](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware)
> > It uses LAN8710A Phy chip.
> > I haven't added any code for handling event, because i'm not sure right now if they're required.
> > Why? Because everything is done automatically by espressif framework.
> > They deal with ESP32 Eth interrupts i guess, because only `ETH.begin()` with proper configuration is needed to establish connection. You can disconnect/reconnect cable, and it will get IP.
> > Moreover, it uses same TCP/IP stack as WiFi, so there's even no restart of listening socket required.
> > Right now i'm trying to include Eth Phy configuration (MDIO,MCS, Power GPIO selection, Phy type selection etc.) into ESPEasy settings, but i cant find place which is executed after submit is pressed on Settings page?
> 
> From Olimex product page :
> 
> ",,,The Si3402-B chip is IEEE 802.3-compliant, including pre-standard (legacy) PoE support. ..."
> 
> Can you confirm the PoE part operates on the non standard (aka passive) PoE 12-48v ?

Sorry i haven't bought an PoE adapter yet :(
But i hope that this device support most of PoE switches/adapter on the market...

ironataerial commented:
> Sorry i haven't bought an PoE adapter yet :(
> But i hope that this device support most of PoE switches/adapter on the market...

802.3af/at PoE does some "handshaking" prior to powering the port and it is 48v. Most of those devices come with a price tag. 
Passive PoE does no checks and shoots power straight up but at a much lower cost on the "injector" part

[https://www.aerial.net/shop/product/1793/](url)

Can use that with any 12v Power Supply.

Having to use a 802.3af/at injector or switch just to power a low cost ESP defeats objectives.

mobrembski committed:
Added Olimex ESP32-PoE board to DeviceModel and default settings

TD-er commented on src/ESPEasy_buildinfo.h:
```diff
@@ -20,7 +20,7 @@
 
 
-#define BUILD                           20104    // git version e.g. "20103" can be read as "2.1.03" (stored in int16_t)
+#define BUILD                           20105    // git version e.g. "20103" can be read as "2.1.03" (stored in int16_t)
```
Are you sure it isn't already set to 20105?
Are you up-to-date with the current HEAD of mega branch?

mobrembski replied:
Yes, i'm 100% sure.
Maybe you haven't pushed some changes?

TD-er replied:
Just given the fact we have this merge conflict, I am sure you're not basing your PR on the latest HEAD of mega:
```
<<<<<<< feature/ethernet_support
#ifdef HAS_ETHERNET
  const unsigned int SettingsStructSize = (252 + 82 * TASKS_MAX);
#else
  const unsigned int SettingsStructSize = (244 + 82 * TASKS_MAX);
#endif
=======
  const unsigned int SettingsStructSize = (248 + 82 * TASKS_MAX);
>>>>>>> mega
```

TD-er replied:
I just resolved this merge conflict, so you should now try to fetch the updated PR.
N.B. there is also some code referring the 20105 build (as `< 20105` ) to initialize newly added members to the settings struct.

mobrembski replied:
Damn, i forget i was fetching origin not upstream..Sorry, and thanks for help :)

mobrembski replied:
I've rebased my branch, let's keep linear history without merges...

TD-er replied:
The reason I did fix the merge conflict was because there have been additions to the SettingsStruct, so that's one thing we really should make sure it won't get lost in merges.

mobrembski committed:
Increased BUILD number to 20107

mobrembski committed:
Fix build on devices without Ethernet.

Forget to adjust size of GpioFactorySettingsStruct.

TD-er commented on src/ESPEasy_checks.ino:
```diff
@@ -58,7 +58,12 @@ void run_compiletime_checks() {
+#ifdef HAS_ETHERNET
+  const unsigned int GpioFactorySettingsStructSize = 17;
+#else
+  const unsigned int GpioFactorySettingsStructSize = 11;
```
Not sure if this is practical.
If switching between builds with and without ethernet, then the settings file will not be exchangable.
The same for the SettingsStruct.
Just leave the member variables in the settings file, or else we will run into issues later when new members are added.

mobrembski replied:
Ok :)

mobrembski committed:
Remove ifdefs for Ethernet in settings

TD-er commented on platformio_esp32_envs.ini:
```diff
@@ -28,7 +28,7 @@ build_flags               = ${mqtt_flags.build_flags}
 extends                   = esp32_common
 platform                  = ${esp32_common.platform}
-build_flags               = ${esp32_common.build_flags}   -DPLUGIN_BUILD_CUSTOM
+build_flags               = ${esp32_common.build_flags}   -DPLUGIN_BUILD_CUSTOM -DESP32 -DHAS_ETHERNET
```
You should not need to define `ESP32` here.
Do you have an example where it is needed?

mobrembski replied:
Yes, i've accidentialy left it here, need to be removed.
BTW: For what this preprocessor define is used for?
I can't find any place where it is defined, none of envs include this

TD-er replied:
ESP32 is defined in the core libs if I'm not mistaken.

TD-er commented on src/ESPEasyEth.ino:
```diff
@@ -0,0 +1,90 @@
+  hostnameToReturn.replace(" ", "-");
+  hostnameToReturn.replace("_", "-"); // See RFC952
+  return hostnameToReturn;
+}
```
Why do you need this extra function? It looks like a function we already have to replace chars in a hostname.

mobrembski replied:
We do have such a function?
Which one? Only i can found was in ESPEasyWiFi.ino but it was strictly related to WiFi

TD-er replied:
Well the same rules apply to the allowed characters.
So if it is no longer limited to WiFi, then move it to Networking.ino.
Better not do copy/paste of code or else people here may confuse my allergy for code duplication with another virus outbreak...

TD-er commented on src/ESPEasyEth.ino:
```diff
@@ -0,0 +1,90 @@
+  return hostnameToReturn;
+}
+
+bool checkSettings() {
```
This is a rather generic function name.
Keep in mind this will be a global defined function.

mobrembski replied:
Fixed

TD-er commented on src/ESPEasyEth.ino:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+void printSettings()
```
Same here, it is a global defined function, with a very generic function name.

mobrembski replied:
Fixed

TD-er commented on src/ESPEasyEth_ProcessEvent.ino:
```diff
@@ -0,0 +1,426 @@
+#include "src/Globals/ESPEasyWiFiEvent.h"
+
+/*
+bool unprocessedWifiEvents() {
```
All in comments and copy paste of the same functions in another file.
If you don't use it, please remove it.
I do use the global search a lot and these functions will then also show up in the search results even when in comments.

mobrembski replied:
Sorry, i've mistaken added this file, it should be removed

TD-er commented on src/WebServer.ino:
```diff
@@ -218,6 +218,7 @@ void WebServerInit()
   web_server.on(F("/i2cscanner"),    handle_i2cscanner);
   #endif
   web_server.on(F("/json"),          handle_json);     // Also part of WEBSERVER_NEW_UI
+  web_server.on(F("/csv"),           handle_csvval);
```
This is part of another PR, which is not (yet) merged.

mobrembski replied:
That's correct, i've branched from that PR for my convenience. 
I hope that PR will be merged too, but in meantime i can rebase it on current mega if you really want.

TD-er commented on src/WebServer_JSON.ino:
```diff
@@ -3,7 +3,75 @@
+// ********************************************************************************
+// Web Interface get CSV value from task
+// ********************************************************************************
+void handle_csvval()
```
This is also of another PR, which is not (yet) merged.

uzi18 commented:
Si3402-B datasheet - minimum voltage 2.8V max 57V

@mobrembski maybe you can try to use 12VDC power adaptor on PoE to check?

mobrembski committed:
Renamed functions in ESPEasyEth, added eth* prefix

mobrembski commented:
> Si3402-B datasheet - minimum voltage 2.8V max 57V
> 
> @mobrembski maybe you can try to use 12VDC power adaptor on PoE to check?

Unfortunately i don't have any PoE adapters :( i will buy one in future.
But i believe it will work, if it supports any PoE standard. I'm not quite sure if it will work with passive PoE. But at least i'm sure that passive POE injectors won't hurt/damange this board.

uzi18 commented:
You can just connect power to blue/brown pairs, recrifier is integrated into ethernet transformer.
I also think it should work

TD-er commented:
802.3af/at do negotiate stuff, so Passive PoE may not always be compatible.
I do have some of those network switches, so I will make sure to order more than one of these modules to test :)

ironataerial commented:
> @mobrembski

Let me have your address. I can mail you a couple passive PoE injectors to test.

mobrembski commented:
> > @mobrembski
> 
> Let me have your address. I can mail you a couple passive PoE injectors to test.

Please send me an email - byku@byku.com.pl - i can;t find neither you address, nor "private message" option in github.com

mobrembski committed:
Moved WifiGetHostname into more generic createRFCCompliantHostname

uzi18 commented:
@TD-er if we can use external file like "ext_eth.dat" for these 6 bytes so we are backward compatible, why we need these settings in esp8266 configs?

TD-er commented:
> @TD-er if we can use external file like "ext_eth.dat" for these 6 bytes so we are backward compatible, why we need these settings in esp8266 configs?

That's a good question.
We can also have esp32 only stuff wrapped in `#ifdef ESP32`, as we also have for some other parameters.
After all the settings files are not compatible between ESP8266 and ESP32.
So maybe we should also rename the settings file for ESP32 to make it clear and then we can also fix some issues like max. nr of GPIO pins to setup for the boot states. (is now limited to nr. of pins for ESP8266)

uzi18 commented:
@TD-er good point to start from :)

mobrembski commented:
> Si3402-B datasheet - minimum voltage 2.8V max 57V
> 
> @mobrembski maybe you can try to use 12VDC power adaptor on PoE to check?

Big Thanks to the @ironataerial - he send me a sample of Passive PoE injector.
So i've have a chance to test this board with passive PoE - unfortunately, it didn't work :(
Tried with 19V DC power supply.
Apparently, Si3402-B is doing negotiation before applying power. I just wondered, AFAIK it does only resistance checking between data lines. Maybe it's possible to create very cheap fake PoE compliant adapter for this?

uzi18 commented:
What if just connections are different?
If you power it with 19VDC than what voltage you can measure on Vpos and  Vneg around chip?

mobrembski commented:
> What if just connections are different?
> If you power it with 19VDC than what voltage you can measure on Vpos and Vneg around chip?

I don't think they're different.
Looking at what i've received from @ironataerial :
https://www.aerial.net/shop/product/1739/
And schematics of Olimex board (ethernet Part, look for Spare1 and Spare2):
https://github.com/OLIMEX/ESP32-POE/blob/master/HARDWARE/ESP32-PoE-hardware-revision-D/ESP32-PoE_Rev_D.pdf
They use the same pairs in RJ45 connector

But sure, i will check it later, when i've get a multimeter

TD-er commented:
As this is not yet going to be merged, but I do need to increase the build value for another setting I need to add.
So can you increase the build nr in your PR to 20107, so I can use 20106?

uzi18 commented:
@TD-er maybe you can also move all esp32 settings to another file in one move?

TD-er commented:
> @TD-er maybe you can also move all esp32 settings to another file in one move?

Well I'm now working on moving all MQTT related controller settings into a separate file, so we can allow for longer user credentials and move the global MQTT settings over there.

So one big change at a time.

mobrembski commented:
Ok, let's discuss what should be done to get this merged (and my second PR too)?
So i can plan work in some way...

> So can you increase the build nr in your PR to 20107, so I can use 20106?

Sure :)

TD-er commented on src/WebServer_HardwarePage.ino:
```diff
@@ -75,6 +83,23 @@ void handle_hardware() {
+  String ethPhyTypes[2] = { F("LAN8710"), F("TLK110") };
+  addSelector("ethtype", 2, ethPhyTypes, NULL, NULL, Settings.ETH_Phy_Type, false, true);
+  addFormNumericBox(F("Ethernet PHY Address"), "ethphy", Settings.ETH_Phy_Addr, 0, 255);
+  addFormNote(F("I²C-address of Ethernet PHY (0 or 1 for LAN8720, 31 for TLK110)"));
```
This looks like non-ASCII, which will give issues on compiling.

mobrembski replied:
I don't expect any issues in neither displaying I2C nor compiling by PlatformIO :)

TD-er replied:
> I don't expect any issues in neither displaying I2C nor compiling by PlatformIO :)

Well the thing is I have seen issues with non-standard ASCII characters in our code (even in comments)
Not sure whether different Git clients or their settings may cause issues, but the way how Arduino code is parsed is not always handling these correct.
PlatformIO should now show warnings when parsing code with non-ASCII characters, which was added when I had those issues last time.
Displaying is another thing. 
When you use characters from the extended ASCII table and not Unicode encoding, it depends on the used code table whether it will be displayed like you intended.
For example in countries with Cyrillic characters I can imagine they have something other than the 'superscript 2' for that character.

If you browse through the code, you can see all special characters are handled via either HTML-encoded Unicode, or HTML-escaped characters.

mobrembski replied:
Ahh, now i understood your intention. Ok, i've changed it to HTML entity &sup2; ;)
Should be even more readable than HTML coded character ;)

TD-er commented:
> Ok, let's discuss what should be done to get this merged (and my second PR too)?
> So i can plan work in some way...

Your PR will have merge issues with mine, so I suggest we will merge mine ( #2972 ) so I suggest we try to merge that one today and yours tomorrow.
I don't see where this PR should hurt existing installs, so I don't see why it could not be merged.

mobrembski committed:
Added basic implementation of Ethernet Static IP

qiudecai commented:
Ethernet can only access the web server, but cannot send data to the controller

uzi18 commented:
@qiudecai even when you disconnect from wifi?

TD-er commented:
I think we may need to add a wrapper object to which all network related code can talk.
Now the use of `WiFiClient` is all over the place in the code.
If we need member functions present in `WiFiClient`, but not in `Client`, then we need such a wrapper class. (or special handling if it is only used in a few places)

I suggest to first try to replace all instances of `WiFiClient` into `Client` and see whether the compiler agrees.
Then we may need some factory design pattern to create the appropriate client where needed (every place where we now create a `WiFiClient` )

This also implies there can only be one active at the same time, so it may be tricky to have support for an AP mode and an ethernet adapter. (not impossible)
But WiFi in STA mode (or STA+AP) and ethernet active at the same time is out of the question. That simply takes way, way, way too much changes in the code.

TD-er commented:
If there is not much collision with other parts of the code, we can decide to merge this now and then try to fix it, so we don't drift away too much from the current code base.  (I am looking at how to improve the initial wifi setup right now)

kretzp commented:
> Ethernet can only access the web server, but cannot send data to the controller

Yes, I saw this, too. I made a fix to it. It is even when you disconnect to the internet. I will commit ist soon.

kretzp commented:
I think I have it up an running that Ethernets works when you disconnected from Wifi. Mqtt is workong as well. A wrapper objetc around Wifi is not needed in my opinion, I have it up an running locally and I will commit it after foxing some code smells.

kretzp committed:
Extended the great work from Michał Obrembski mobrembski, so that:
- mqtt works
- WifiConnected works as EthernetConnected

kretzp commented:
You could take a look at my work here if you want:
https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support

kretzp commented:
I have an Olimex-EVB board and it works well:
https://www.olimex.com/Products/IoT/ESP32/ESP32-EVB/open-source-hardware

TD-er commented:
I just received my LAN8720 boards, so I should be able to setup a test here. (now find some time for it...)

TD-er commented:
> You could take a look at my work here if you want:
> https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support

I looked through the last 2 commits and I don't see things that immediately catch my eye.
So that looks OK.
What I am missing is the change that made the controller work. But maybe that has already been dealt with in the Arduino Core lib.
We use `WiFiClient` in a lot of our code, so I wonder what we're missing when keep using that class while ETH is configured.

There is a merge conflict with `ESPEasyDefaults.h` can you also fix that (or I can do it) so we can see what Travis has to say about the PR.

kretzp commented:
I got a merge conflict there, too when I pulled the branch. Could you resolve it please?
I think it should be the like this:
#ifndef DEFAULT_PIN_RESET_BUTTON
#define DEFAULT_PIN_RESET_BUTTON         (-1)

#endif
#ifndef DEFAULT_ETH_PHY_ADDR
#define DEFAULT_ETH_PHY_ADDR             0
#endif
#ifndef DEFAULT_ETH_PHY_TYPE
#define DEFAULT_ETH_PHY_TYPE             0
#endif
#ifndef DEFAULT_ETH_PIN_MDC
#define DEFAULT_ETH_PIN_MDC              23
#endif
#ifndef DEFAULT_ETH_PIN_MDIO
#define DEFAULT_ETH_PIN_MDIO             18
#endif
#ifndef DEFAULT_ETH_PIN_POWER
#define DEFAULT_ETH_PIN_POWER            -1
#endif
#ifndef DEFAULT_ETH_CLOCK_MODE
#define DEFAULT_ETH_CLOCK_MODE           3
#endif

kretzp commented:
> What I am missing is the change that made the controller work. But maybe that has already been dealt with in the Arduino Core lib.
> We use `WiFiClient` in a lot of our code, so I wonder what we're missing when keep using that class while ETH is configured.

You could take a look at
https://github.com/espressif/arduino-esp32/blob/master/libraries/WiFi/examples/ETH_LAN8720/ETH_LAN8720.ino

Her you see how ETH.h works with WifiClient. It works well and I remember the connected state like in this example in
static bool eth_connected = false;

TD-er commented:
> static bool eth_connected = false;

That's tricky as you must be sure that flag is also updated on each connect/disconnect.
I will have a look at the merge conflict and that part of the code you mentioned.

TD-er committed:
Merge branch 'mega' into feature/ethernet_support

kretzp commented:
> > static bool eth_connected = false;
> 
> That's tricky as you must be sure that flag is also updated on each connect/disconnect.
> I will have a look at the merge conflict and that part of the code you mentioned.

Yes I know, but I think this is done via event handling in function
void ETHEvent(WiFiEvent_t event)
    case SYSTEM_EVENT_ETH_DISCONNECTED:
      addLog(LOG_LEVEL_ERROR, F("ETH Disconnected"));
      eth_connected = false;
      break;
    case SYSTEM_EVENT_ETH_STOP:
      addLog(LOG_LEVEL_INFO, F("ETH Stopped"));
      eth_connected = false;
      break;

kretzp commented:
I tested this by pulling an pushing the ethernet cable.
It works fine and you'll see the events in the logs.

mobrembski commented:
> I tested this by pulling an pushing the ethernet cable.
> It works fine and you'll see the events in the logs.

Thanks very much for you work! :)
I have a question did you check EspEasy P2P networking ?

kretzp commented:
> > I tested this by pulling an pushing the ethernet cable.
> > It works fine and you'll see the events in the logs.
> 
> Thanks very much for you work! :)
> I have a question did you check EspEasy P2P networking ?

No I didn't check it. But I alway wanted to try ESPEasy P2p networking. Perhaps I will check it this evening.

mobrembski committed:
Removed FEATURE_MDNS from pre_custom_esp82xx.py

It looks like a bug, esp82xx doesn't support MDNS due to missing WifiGetHostname.
Seems that it was accidentially added in a3a10f2

ironataerial commented:
> No I didn't check it. But I alway wanted to try ESPEasy P2p networking. Perhaps I will check it this evening.

It works over WiFi but not over ETH. Make sure you disable WiFi first(I entered bogus credentials) prior to enabling P2P . It takes some time to time out of the listing but even so will be in the WiFi IP, just so you will not get the wrong impression.

ironataerial commented:
Which also raises the issue of being able to properly disable the wless iface for ETH only operation.

TD-er commented on platformio_esp32_envs.ini:
```diff
@@ -52,4 +52,31 @@ debug_tool                = ftdi
+upload_protocol           = ftdi
+debug_tool                = ftdi
+debug_extra_cmds          = break Misc.ino:3011
+extra_scripts             = ${esp32_common.extra_scripts}
```
These environments can be much more simple.
The only thing you want to change is the `build_flags`, so then do that.

For example:

```ini
[env:test_ESP32-wrover-kit_4M316k_ETH]
extends                   = test_ESP32-wrover-kit_4M316k
build_flags               = ${test_ESP32-wrover-kit_4M316k.build_flags}  -DHAS_ETHERNET
```

kretzp replied:
Yes, thats right of course, I change it.

kretzp replied:
Didn't get it. When I do these changes, I do not see the build option in platformio anymore

TD-er replied:
How did you change it?

kretzp replied:
This didn't work
[env:test_ESP32-wrover-kit_4M316k_ETH]
extends                   = env:test_ESP32-wrover-kit_4M316k
build_flags               = ${env:test_ESP32-wrover-kit_4M316k.build_flags} -DHAS_ETHERNET

This works
[env:test_ESP32-wrover-kit_4M316k_ETH]
extends                   = env:test_ESP32-wrover-kit_4M316k
platform                  = ${env:test_ESP32-wrover-kit_4M316k.platform}
build_flags               = ${env:test_ESP32-wrover-kit_4M316k.build_flags} -DHAS_ETHERNET

It seems, tha platform is mandatory. I don't know why. I changed it and pushed it to my branch
https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support

TD-er commented on src/ESPEasy.ino:
```diff
@@ -1012,3 +1021,45 @@ void backgroundtasks()
 }
+
+#ifdef HAS_ETHERNET
+void ETHEvent(WiFiEvent_t event)
```
Similar function for WiFi also has its own .ino file
So please move this to some file named `Eth_events.ino` or maybe a more appropriate file name.

kretzp replied:
I will move this to
ESPEasy**ETH**Event.ino as it is called ESPEasy**Wifi**Event.ini in Wifi case.

kretzp replied:
Change pushed to https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support

TD-er commented on src/ESPEasyWifi.ino:
```diff
@@ -75,6 +75,10 @@
 
+  #ifdef HAS_ETHERNET
+  return eth_connected;
+  #endif
```
This will give a compiler warning or maybe error when `HAS_ETHERNET` is defined.
There is some "unreachable code" after this statement.

kretzp replied:
I will change it to
`#ifdef HAS_ETHERNET
  return eth_connected;
#else
....
  #endif`

kretzp replied:
Change pushed to https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support

kretzp commented:
> Which also raises the issue of being able to properly disable the wless iface for ETH only operation.

When you compile with flag HAS_ETHERNET, it should be disabled. No Wifi will be started at all.

kretzp commented:
> > No I didn't check it. But I alway wanted to try ESPEasy P2p networking. Perhaps I will check it this evening.
> 
> It works over WiFi but not over ETH. Make sure you disable WiFi first(I entered bogus credentials) prior to enabling P2P . It takes some time to time out of the listing but even so will be in the WiFi IP, just so you will not get the wrong impression.

I tested it and it did not work. I made some minor changes and pushed them to https://github.com/kretzp/ESPEasy/tree/feature/ethernet_support
Now it works well.

kretzp commented:
> I think we may need to add a wrapper object to which all network related code can talk.
> Now the use of `WiFiClient` is all over the place in the code.
> If we need member functions present in `WiFiClient`, but not in `Client`, then we need such a wrapper class. (or special handling if it is only used in a few places)
> 
> I suggest to first try to replace all instances of `WiFiClient` into `Client` and see whether the compiler agrees.
> Then we may need some factory design pattern to create the appropriate client where needed (every place where we now create a `WiFiClient` )
> 
> This also implies there can only be one active at the same time, so it may be tricky to have support for an AP mode and an ethernet adapter. (not impossible)
> But WiFi in STA mode (or STA+AP) and ethernet active at the same time is out of the question. That simply takes way, way, way too much changes in the code.

I walked a lot through the code yesterday, and now my opinion is, to would be good to have such a wrapper. So I propose, that I will make a branch and will code the following:
- There will be a configuration in "Hardware" if you want to use ethernet or wifi, not both at the same time, so you can change it, without rebuild, just configuration.
- The compile parameter HAS_ETHERNET will be there to compile the source code with additional ethernet support
- All networking should be done via networking.ino, which will be the wrapper. For example you call Network.localIp() and you'll get the IP address in wifi or ethernet case. The handling will be done within networking.ino and it will call ETH.localIp() or Wifi.localIp() depending on operation mode ehternet or wifi.

kretzp commented:
> > I think we may need to add a wrapper object to which all network related code can talk.
> > Now the use of `WiFiClient` is all over the place in the code.
> > If we need member functions present in `WiFiClient`, but not in `Client`, then we need such a wrapper class. (or special handling if it is only used in a few places)
> > I suggest to first try to replace all instances of `WiFiClient` into `Client` and see whether the compiler agrees.
> > Then we may need some factory design pattern to create the appropriate client where needed (every place where we now create a `WiFiClient` )
> > This also implies there can only be one active at the same time, so it may be tricky to have support for an AP mode and an ethernet adapter. (not impossible)
> > But WiFi in STA mode (or STA+AP) and ethernet active at the same time is out of the question. That simply takes way, way, way too much changes in the code.
> 
> I walked a lot through the code yesterday, and now my opinion is, to would be good to have such a wrapper. So I propose, that I will make a branch and will code the following:
> 
> * There will be a configuration in "Hardware" if you want to use ethernet or wifi, not both at the same time, so you can change it, without rebuild, just configuration.
> * The compile parameter HAS_ETHERNET will be there to compile the source code with additional ethernet support
> * All networking should be done via networking.ino, which will be the wrapper. For example you call Network.localIp() and you'll get the IP address in wifi or ethernet case. The handling will be done within networking.ino and it will call ETH.localIp() or Wifi.localIp() depending on operation mode ehternet or wifi.

And another things comes to my mind. When Wifi ist configured, the AP Mode appears like now, when no valid Wifi is found, no change in that.

TD-er commented:
> * All networking should be done via networking.ino, which will be the wrapper.

Preferably make it a .h/.cpp class object.
As I'm moving stuff into those anyway, let's have it right to start with.
The reason for this is that .h/.cpp classes are much better guaranteed to give predictable builds.
Also if classes defined in .ino files must be referred to from other .h/.cpp objects, you must use some extra forward declarations and hope for the best the Arduino way of compiling isn't making its own forward declarations which will be causing a build fail at best or unstable builds at worst.

uzi18 commented:
if you got ethernet cable connection than usually do not need wifi anymore

TD-er commented:
> if you got ethernet cable connection than usually do not need wifi anymore

It can still be useful.
For example if you have had it set to static IP (on LAN), but now switched to another network with another subnet.

Another use case can be to act as a gateway for other nodes, either via WiFi or as ESP-now master.

So I would say it is very useful to have both options available. Not to say they have to be enabled at the same time.

kretzp commented:
> > * All networking should be done via networking.ino, which will be the wrapper.
> 
> Preferably make it a .h/.cpp class object.
> As I'm moving stuff into those anyway, let's have it right to start with.
> The reason for this is that .h/.cpp classes are much better guaranteed to give predictable builds.
> Also if classes defined in .ino files must be referred to from other .h/.cpp objects, you must use some extra forward declarations and hope for the best the Arduino way of compiling isn't making its own forward declarations which will be causing a build fail at best or unstable builds at worst.

OK, no problem, I will do that.

kretzp commented:
> > if you got ethernet cable connection than usually do not need wifi anymore
> 
> It can still be useful.
> For example if you have had it set to static IP (on LAN), but now switched to another network with another subnet.
> 
> Another use case can be to act as a gateway for other nodes, either via WiFi or as ESP-now master.
> 
> So I would say it is very useful to have both options available. Not to say they have to be enabled at the same time.

I propose, to firts start with a configuration to choose between ethernet and wifi in Hardware Configuration via Web GUI. If this works, we can implmenet a "both" option. But then we have to configure for example via which network mqtt has to be done. But this should be the second step.

kretzp commented:
> > > if you got ethernet cable connection than usually do not need wifi anymore
> > 
> > 
> > It can still be useful.
> > For example if you have had it set to static IP (on LAN), but now switched to another network with another subnet.
> > Another use case can be to act as a gateway for other nodes, either via WiFi or as ESP-now master.
> > So I would say it is very useful to have both options available. Not to say they have to be enabled at the same time.
> 
> I propose, to firts start with a configuration to choose between ethernet and wifi in Hardware Configuration via Web GUI. If this works, we can implmenet a "both" option. But then we have to configure for example via which network mqtt has to be done. But this should be the second step.

OK guys, I'm ready for now. I implemented Ethernet or Wifi, not both at same time as first step. Take a look at my pull request:
https://github.com/letscontrolit/ESPEasy/pull/3040
or my branch
https://github.com/kretzp/ESPEasy/tree/feature/ethernet_or_wifi_support
Please test it, I'm looking forward to your comments

TD-er commented:
> Please test it, I'm looking forward to your comments

Today I received my Olimex board, so I should be able to test it this week.

kretzp commented:
> > Please test it, I'm looking forward to your comments
> 
> Today I received my Olimex board, so I should be able to test it this week.

Cool. Which one did you get? I have an ESP32-EVB-EA-IND.

mobrembski commented:
@kretzp 
By the way, i cant build your branch due to:

src/ESPEasyWiFiEvent.cpp: In function 'void WiFiEvent(system_event_id_t, system_event_info_t)':
src/ESPEasyWiFiEvent.cpp:116:50: error: no matching function for call to 'safe_strncpy(char [40], String)'
       safe_strncpy(hostname, NetworkGetHostname()).c_str(), sizeof(hostname));
                                                  ^
In file included from src/ESPEasyWiFiEvent.cpp:10:0:
src/ESPEasy_fdwdecl.h:102:6: note: candidate: bool safe_strncpy(char*, const String&, size_t)
 bool safe_strncpy(char         *dest,
      ^
src/ESPEasy_fdwdecl.h:102:6: note:   candidate expects 3 arguments, 2 provided
src/ESPEasy_fdwdecl.h:105:6: note: candidate: bool safe_strncpy(char*, const char*, size_t)
 bool safe_strncpy(char       *dest,

Some kind of version mismatch?

UPDATE: Ok, wrong parenthesis, fixed that, but please fix that on your branch too :)

TD-er commented:
> > > Please test it, I'm looking forward to your comments
> > 
> > 
> > Today I received my Olimex board, so I should be able to test it this week.
> 
> Cool. Which one did you get? I have an ESP32-EVB-EA-IND.

Yep me too, a very crowded board, so looked like a good test board for ESPEasy testing.

Also ordered a NB-IoT board, so that's for later to play with :)

ironataerial commented:
ESP32_eth device appears in the list of nodes on ESP8266 devices, but only ESP32_eth appears on its own list of nodes.
Sendto does not seem to work for me,
Sendto,5,event,lightson  Works when originated from ESP8266 (with tolerant last parameter enabled), same does not work when originated from ESP32_eth 
-D
![ESP32_eth](https://user-images.githubusercontent.com/11705083/80580402-256b7600-8a14-11ea-9ab1-dc5631fac0cc.jpg)
![ESP8266](https://user-images.githubusercontent.com/11705083/80580406-27cdd000-8a14-11ea-8a49-7c53b4d164bd.jpg)

TD-er commented:
Have you restarted ESPEasy after changing the UDP port number?

Also you're using a strange subnet for 'local' usage.
The 195.a.b.c is also available on the internet as possible IPs, so you may run into other troubles of unreachable hosts online when this is your local network.
It has very likely nothing to do with not being able to communicate via p2p, as that's likely in need of a reboot, but just something you should be aware of.

ironataerial commented:
> Have you restarted ESPEasy after changing the UDP port number?

Prior to the reboot would not register in the node list of the ESP8266s. 

The screenshots and tests were performed after the reboot (both soft and hard)

ironataerial commented:
I use 65500 port and the ESP8266s are on static IPs, if any of these makes any difference

kretzp commented:
> > > > Please test it, I'm looking forward to your comments
> > > 
> > > 
> > > Today I received my Olimex board, so I should be able to test it this week.
> > 
> > 
> > Cool. Which one did you get? I have an ESP32-EVB-EA-IND.
> 
> Yep me too, a very crowded board, so looked like a good test board for ESPEasy testing.
> 
> Also ordered a NB-IoT board, so that's for later to play with :)

That's cool. I will use this board as a pool controller fpr ph an ORP measurment and the relays for injection oh ph fluids and chlorium.
Board works very good in my opinion.

kretzp commented:
> ESP32_eth device appears in the list of nodes on ESP8266 devices, but only ESP32_eth appears on its own list of nodes.
> Sendto does not seem to work for me,
> Sendto,5,event,lightson  Works when originated from ESP8266 (with tolerant last parameter enabled), same does not work when originated from ESP32_eth 
> -D
> ![ESP32_eth](https://user-images.githubusercontent.com/11705083/80580402-256b7600-8a14-11ea-9ab1-dc5631fac0cc.jpg)
> ![ESP8266](https://user-images.githubusercontent.com/11705083/80580406-27cdd000-8a14-11ea-8a49-7c53b4d164bd.jpg)

I will recheck this and will be back this evening.

kretzp commented:
> @kretzp
> By the way, i cant build your branch due to:
> 
> src/ESPEasyWiFiEvent.cpp: In function 'void WiFiEvent(system_event_id_t, system_event_info_t)':
> src/ESPEasyWiFiEvent.cpp:116:50: error: no matching function for call to 'safe_strncpy(char [40], String)'
> safe_strncpy(hostname, NetworkGetHostname()).c_str(), sizeof(hostname));
> ^
> In file included from src/ESPEasyWiFiEvent.cpp:10:0:
> src/ESPEasy_fdwdecl.h:102:6: note: candidate: bool safe_strncpy(char*, const String&, size_t)
> bool safe_strncpy(char _dest, ^ src/ESPEasy_fdwdecl.h:102:6: note: candidate expects 3 arguments, 2 provided src/ESPEasy_fdwdecl.h:105:6: note: candidate: bool safe_strncpy(char_, const char*, size_t)
> bool safe_strncpy(char *dest,
> 
> Some kind of version mismatch?
> 
> UPDATE: Ok, wrong parenthesis, fixed that, but please fix that on your branch too :)

I see it in my sources, too. Strange thing. Since Travis CI reported success:
![image](https://user-images.githubusercontent.com/15065072/80584949-c6593180-8a12-11ea-9d95-104a4f40c2e2.png)
I fixed it in my branch.

kretzp commented:
> > ESP32_eth device appears in the list of nodes on ESP8266 devices, but only ESP32_eth appears on its own list of nodes.
> > Sendto does not seem to work for me,
> > Sendto,5,event,lightson  Works when originated from ESP8266 (with tolerant last parameter enabled), same does not work when originated from ESP32_eth 
> > -D
> > ![ESP32_eth](https://user-images.githubusercontent.com/11705083/80580402-256b7600-8a14-11ea-9ab1-dc5631fac0cc.jpg)
> > ![ESP8266](https://user-images.githubusercontent.com/11705083/80580406-27cdd000-8a14-11ea-8a49-7c53b4d164bd.jpg)
> 
> I will recheck this and will be back this evening.

So I have it up and running. I will just reorder my code and will check it in soon:
![image](https://user-images.githubusercontent.com/15065072/80621046-0a662980-8a47-11ea-9c9f-975b0fa303c4.png)

![image](https://user-images.githubusercontent.com/15065072/80621091-16ea8200-8a47-11ea-8163-6dc13261ce0d.png)

kretzp commented:
> > > ESP32_eth device appears in the list of nodes on ESP8266 devices, but only ESP32_eth appears on its own list of nodes.
> > > Sendto does not seem to work for me,
> > > Sendto,5,event,lightson  Works when originated from ESP8266 (with tolerant last parameter enabled), same does not work when originated from ESP32_eth 
> > > -D
> > > ![ESP32_eth](https://user-images.githubusercontent.com/11705083/80580402-256b7600-8a14-11ea-9ab1-dc5631fac0cc.jpg)
> > > ![ESP8266](https://user-images.githubusercontent.com/11705083/80580406-27cdd000-8a14-11ea-8a49-7c53b4d164bd.jpg)
> > 
> > 
> > I will recheck this and will be back this evening.
> 
> So I have it up and running. I will just reorder my code and will check it in soon:
> ![image](https://user-images.githubusercontent.com/15065072/80621046-0a662980-8a47-11ea-9c9f-975b0fa303c4.png)
> 
> ![image](https://user-images.githubusercontent.com/15065072/80621091-16ea8200-8a47-11ea-8163-6dc13261ce0d.png)

It's checked in in my branch now. Please take a look.
https://github.com/kretzp/ESPEasy/tree/feature/ethernet_or_wifi_support
Perhaps you could comment here:
https://github.com/letscontrolit/ESPEasy/pull/3040
Becaus this is the branch and pull request I am working on. Thanks.

mobrembski commented:
Ok guys, as i can see work on this topic is continued in https://github.com/letscontrolit/ESPEasy/pull/3040
by @kretzp ,
Currently i don't have a time to help :(
So let's close this PR ;)
